Mnemonic Startup GUI

mrmetech opened the pull request:
This PR was for the #451 a bounty Navcoin had to add a Startup GUI which allowed for the ability to restore and create wallets with Mnemonic words. Mnemonic words are like a read readable master key for BIP32 a bitcoin proposal to make key generation derived from one key or a masterkey instead of being randomly created as before hand.

This page is the first page and it allows for the user to select Restore wallet an option to restore a wallet with your mnemonic words or Create a new wallet if you have never created a wallet that supports Mnemonic words before
![image](https://user-images.githubusercontent.com/31669092/74597500-9b6b5e00-501d-11ea-8e84-e5ee62a88d6e.png)

The restore page is where you would enter your mnemonic words if you were restoring it from the mnemonic words you kept safe. If a word is spelt incorrectly or invalid the box will have a red outline, if it is correct it will have a purple outline.
![image](https://user-images.githubusercontent.com/31669092/74597508-d4a3ce00-501d-11ea-97d0-31ba92564946.png)

The reveal page. This page is shown after the user selects Create New Wallet and shows the user his Mnemonic words that he must write down and keep safe.
![image](https://user-images.githubusercontent.com/31669092/74597519-07e65d00-501e-11ea-823d-81c1573dae4f.png)

The sort page. This page is shown after the reveal page and it is used to confirm whether the user has written down his words.
![image](https://user-images.githubusercontent.com/31669092/74597537-3feda000-501e-11ea-82d4-b276ab0a875a.png)


The address the bounty should be paid to NdTrSEe6bMMvVwNUimxs2BiWkX5Xn9MW7o

mxaddict commented:
I'm out of the house right now, will review when I get home.

On Sun, Feb 16, 2020, 09:33 alex v. <notifications@github.com> wrote:

> @aguycalled <https://github.com/aguycalled> requested your review on: #659
> <https://github.com/navcoin/navcoin-core/pull/659> Mnemonic Startup GUI.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/navcoin/navcoin-core/pull/659?email_source=notifications&email_token=AAIDAKPAD2GGS3MKXDMWXUTRDCJWXA5CNFSM4KV6CSV2YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOWVC5LGA#event-3041252760>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDAKK2BB3XIIRIPKY4KBTRDCJWXANCNFSM4KV6CSVQ>
> .
>

chasingkirkjufell commented:
Same as mxaddict.

navbuilder commented:
A new build of 5ae6e0da50a94e4e308cfdade2e72abe9df5122b has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

navbuilder commented:
A new build of 208b4f0216964872ae7d44ec08608402db0f0391 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

chasingkirkjufell commented:
Gitian build of 208b4f0 on Ubuntu 19.04 64GB RAM.

Observation 1:
Wallet client shows welcome page even though wallet.dat was previously created. Users do not have to option to skip the welcome page but only have to choose between creating a new wallet vs importing an old one. 

Creating an wallet when wallet.dat already exists actually displays the mnemonic seeds of the existing wallet instead of a new wallet.

Expectation 1:
Wallet client should not have a welcome page when wallet.dat already exists.

Observation 2:
When wallet.dat does not exist, after re-entering all the mnemonic seeds, the wallet will fail to boot up and show 
```key.cpp:148: CPubKey CKey::GetPubKey() const: Assertion `fValid' failed.```

Expectation 2:
When wallet.dat does not exist, after re-entering all the mnemonic seeds, the wallet should load normally.

Observation 3 (non-critical):
When re-entering the seeds, the only way to correct a misplaced seed word is to drag and drop another one on top of it. It'd be nice if the users can simply drag and drop the word back down. An extreme case would be if a user finds out the seeds are misplaced after using all the words, the user will have no way to correct it but return to the previous step and come back doing it all over again.

Also tested:
Importing devnet wallet worked without issue.

mrmetech commented:
> Observation 1:
> Wallet client shows welcome page even though wallet.dat was previously created. Users do not have to option to skip the welcome page but only have to choose between creating a new wallet vs importing an old one.

I have now fixed this issue stated above ^

> Observation 2:
> When wallet.dat does not exist, after re-entering all the mnemonic seeds, the wallet will fail to boot up and show
> `` key.cpp:148: CPubKey CKey::GetPubKey() const: Assertion `fValid' failed. ``

I have also just fixed this issue stated above the issue was that I was redefining words in wallet.cpp instead of just setting the value.

> Observation 3 (non-critical):
> When re-entering the seeds, the only way to correct a misplaced seed word is to drag and drop another one on top of it. It'd be nice if the users can simply drag and drop the word back down. An extreme case would be if a user finds out the seeds are misplaced after using all the words, the user will have no way to correct it but return to the previous step and come back doing it all over again.

For Observation 3 that is just an inherent issue with Qt drag and drop system and nothing can really be done about that. I spent hours looking into that 8 months back.

Recap Both issues 1&2 were fixed and then I squashed the commit, also issue 3 is just an issue with how qt works and that is the best solution for drag and drop I found.

navbuilder commented:
A new build of 59a69a21dc60fa0a958b95cd9e03e1fb146a6a48 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

navbuilder commented:
A new build of f1cfe1ac611c7a36729ed1cd51f75255501423ad has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

aguycalled commented:
<img width="809" alt="image" src="https://user-images.githubusercontent.com/24814046/74603372-2480a480-50b3-11ea-9aed-8cf36ce3e93b.png">

When I click in one of the words it is highlighted. If some words in other columns are also highlighted, those are not reset.

aguycalled commented:
<img width="807" alt="image" src="https://user-images.githubusercontent.com/24814046/74603403-7e816a00-50b3-11ea-8c89-7df0ffd67ea9.png">
This window looks too big for the little content and too dry. Would it be possible to make it smaller and maybe add the NavCoin logo from the splash screen?

mxaddict commented:
GUI is asking me to restore/create a key when I already have a wallet.dat

![Capture](https://user-images.githubusercontent.com/1060905/74607415-c46e1a80-5113-11ea-8273-a199d4e66e23.PNG)

mxaddict commented:
I would suggest asking for 8 words randomly instead of all 24 words (This would make it easier for the user)

For example have 16 of the words already pre filled on the drag and drop stage.

mxaddict commented:
Also might be better if instead of drag and drop, it would be just click the words on the order that they are required (Drag n Drop is harder then just clicking the correct words in the correct order)

And if they make a mistake, they can just click the word that is wrong (In the top section) to remove it.

mxaddict commented:
I also see some of the styles are still inline, I can help move these into a stylesheet (Once the GUI PR #557 is merged)

Other than that, nice work on the PR.

mrmetech commented:
> GUI is asking me to restore/create a key when I already have a wallet.dat

I  Already fixed that can you can confirm which commit you compiled the current commit is: f1cfe1ac611c7a36729ed1cd51f75255501423ad

mrmetech commented:
> I would suggest asking for 8 words randomly instead of all 24 words (This would make it easier for the user)
> 
> For example have 16 of the words already pre filled on the drag and drop stage.

This is a onetime setup, so it shouldn't be that much of a hassle do all 24. If the user had to do this more then once in his lifetime, that would be more of a concern, but the goal he knows the placement of all 24 words.

mxaddict commented:
> > GUI is asking me to restore/create a key when I already have a wallet.dat
> 
> I Already fixed that can you can confirm which commit you compiled the current commit is: [f1cfe1a](https://github.com/navcoin/navcoin-core/commit/f1cfe1ac611c7a36729ed1cd51f75255501423ad)

I cloned a fresh copy just a  an hour ago, so it was latest commit for sure.

Tested on Ubuntu LTS 18.04

mxaddict commented:
> > I would suggest asking for 8 words randomly instead of all 24 words (This would make it easier for the user)
> > For example have 16 of the words already pre filled on the drag and drop stage.
> 
> This is a onetime setup, so it shouldn't be that much of a hassle do all 24. If the user had to do this more then once in his lifetime, that would be more of a concern, but the goal he knows the placement of all 24 words.

This is exactly why I suggested, I do this kind of thing (Importing keys, don't ask why 😅 ) about 1-2 times a month 👍

mxaddict commented:
> > > GUI is asking me to restore/create a key when I already have a wallet.dat
> > 
> > 
> > I Already fixed that can you can confirm which commit you compiled the current commit is: [f1cfe1a](https://github.com/navcoin/navcoin-core/commit/f1cfe1ac611c7a36729ed1cd51f75255501423ad)
> 
> I cloned a fresh copy just a an hour ago, so it was latest commit for sure.
> 
> Tested on Ubuntu LTS 18.04

I'm trying the windows version now.

mrmetech commented:
> Also might be better if instead of drag and drop, it would be just click the words on the order that they are required (Drag n Drop is harder then just clicking the correct words in the correct order)

^ While that sounds like a good idea, that would probably require a rewrite of what I wrote which wouldn't really be possible at this time

> And if they make a mistake, they can just click the word that is wrong (In the top section) to remove it.

^ This is a pretty good idea I can probably add that, the only issue would be which QGraphicsList it got put into which other then that it is possible.

mxaddict commented:
> > Also might be better if instead of drag and drop, it would be just click the words on the order that they are required (Drag n Drop is harder then just clicking the correct words in the correct order)
> 
> ^ While that sounds like a good idea, that would probably require a rewrite of what I wrote which wouldn't really be possible at this time
> 
> > And if they make a mistake, they can just click the word that is wrong (In the top section) to remove it.
> 
> ^ This is a pretty good idea I can probably add that, the only issue would be which QGraphicsList it got put into which other then that it is possible.

Not really deal breakers, I think in the current design we can merge these changes and make modifications to it as needed, important thing is we have a GUI for non CLI savy users to import the keys, good work.

mrmetech commented:
> When I click in one of the words it is highlighted. If some words in other columns are also highlighted, those are not reset.

In regards to that, I will address that by make it so if you select one its color won't change only when you hover over it.

mxaddict commented:
> > > > GUI is asking me to restore/create a key when I already have a wallet.dat
> > > 
> > > 
> > > I Already fixed that can you can confirm which commit you compiled the current commit is: [f1cfe1a](https://github.com/navcoin/navcoin-core/commit/f1cfe1ac611c7a36729ed1cd51f75255501423ad)
> > 
> > 
> > I cloned a fresh copy just a an hour ago, so it was latest commit for sure.
> > Tested on Ubuntu LTS 18.04
> 
> I'm trying the windows version now.

I've tested windows version, and seems to work as intended, so the issue for it not detecting the wallet.dat persists on linux build.

I'll poke around and see if I can figure out what is the cause.

mxaddict commented:
Build: f1cfe1a
Windows 10 64bit
- [ ] Restore works
- [ ] Creating new key works

mxaddict commented:
Build: f1cfe1a
Ubuntu LTS 18.04
- [x] Restore works
- [x] Creating new key works
- [ ] Does not ask to import/create if already have a wallet.dat file

aguycalled commented on src/mnemonic/generateseed.cpp:
```diff
@@ -0,0 +1,41 @@
+    CKey key;
+    key.MakeNewKey(true);
+
+    std::vector<unsigned char> keyData;
```
using this vector to store the key data is unsafe, as it will leave traces in memory of the seed.
an option would be to modify create_mnemonic to take directly a CKey object, which uses secure_allocator (src/support/allocators/secure.h) and ensures memory is clean when the object is destroyed

mrmetech replied:
This is just code from Veil you didn't pull to Navcoin when bringing over Mnemonic to Navcoin so I would really have no idea what is best.

mrmetech replied:
Ok that is fixed now

mxaddict commented:
Build: f795e44
Windows 10 and Ubuntu 18.04

- [x] Restore works
- [x] Creating new key works
- [x] Does not ask to import/create if already have a wallet.dat file on this network (mainnet, devnet, testnet)

navbuilder commented:
A new build of b539b58c20e15b608377def1ee7c6c23f897201c has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

navbuilder commented:
A new build of f795e44a816f27a465911aa6b14e4466443b876f has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

chasingkirkjufell commented:
Build: f795e44
Ubuntu 19.04

- [x] Restore works
- [x] Creating new key works
- [x] Does not ask to import/create if already have a wallet.dat file on this network (mainnet, devnet, testnet)

I am happy with this PR now. I will leave the final review for @aguycalled for OSX system.

aguycalled commented on src/qt/startoptionsmain.cpp:
```diff
@@ -0,0 +1,238 @@
+    : QDialog(parent), ui(new Ui::StartOptionsMain) {
+    ui->setupUi(this);
+
+    this->setWindowTitle("Navcoin Wallet Setup");
```
this string should go through _() or tr() in order to be translatable

aguycalled commented on src/qt/startoptionsmain.cpp:
```diff
@@ -0,0 +1,238 @@
+            rtrim(words_empty_str);
+            rtrim(words_mnemonic);
+            if (words_empty_str != words_mnemonic) {
+                QString error = "Unfortunately, your words are in the wrong "
```
this string should go through _() or tr() in order to be translatable

aguycalled commented on src/qt/startoptionsmain.cpp:
```diff
@@ -0,0 +1,238 @@
+                QApplication::quit();
+            } else {
+                QString error =
+                    "Unfortunately, your words seem to be invalid. This is "
```
this string should go through _() or tr() in order to be translatable

aguycalled commented:
in OSX 10.14 the list of words moves upwards when dragging words, covering the last row of cells

<img width="818" alt="image" src="https://user-images.githubusercontent.com/24814046/74641339-6b34d400-5171-11ea-93da-725bc71269f8.png">

aguycalled commented on src/qt/navcoin.cpp:
```diff
@@ -333,13 +339,35 @@ void NavCoinApplication::createOptionsModel(bool resetSettings)
+    }
+
+    std::string walletFile = GetArg("-wallet", "wallet.dat");
+    if (fs::exists(walletFile)) return true;
```
Could we add here...

```c++
|| GetBoolArg("-skipmnemonicbackup",false)
```
?

mrmetech commented:
> in OSX 10.14 the list of words covers the last row of cells
> 
> <img alt="image" width="818" src="https://user-images.githubusercontent.com/24814046/74641339-6b34d400-5171-11ea-93da-725bc71269f8.png">

Can you try making the window larger for context of what will happen

chasingkirkjufell commented:
I'm seeing it on Ubuntu too now. The last row will show up if the window is enlarged.

mxaddict commented:
I think it's caused the the extra text we added to the top.

On Tue, Feb 18, 2020, 00:23 chasingkirkjufell <notifications@github.com>
wrote:

> I'm seeing it on Ubuntu too now. The last row will show up if the window
> is enlarged.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/navcoin/navcoin-core/pull/659?email_source=notifications&email_token=AAIDAKKAACM3WCAJCNCQ7GTRDK2XLA5CNFSM4KV6CSV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL67VEQ#issuecomment-587070098>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDAKI37MGKBI4JNNSAZKDRDK2XLANCNFSM4KV6CSVQ>
> .
>

mrmetech commented:
I made a change that I believe will fix the issue found of linux and mac, If someone could test it and take a screenshot for me thanks.

chasingkirkjufell commented on src/qt/startoptionssort.cpp:
```diff
@@ -0,0 +1,187 @@
+        view = new QGraphicsView;
+        view->setScene(scene);
+        view->setMinimumSize(QSize(1, 250));
+        view->setContentsMargins(0, 0, 0, 2)
```
missing ;

chasingkirkjufell commented:
4th row now correctly shown on commit a2fd710.
![image](https://user-images.githubusercontent.com/38635377/74680079-40547b00-5185-11ea-98f5-81a65f9210c2.png)

mxaddict commented:
Nice <3

On Tue, Feb 18, 2020, 02:59 chasingkirkjufell <notifications@github.com>
wrote:

> 4th row now correctly shown on commit a2fd710
> <https://github.com/navcoin/navcoin-core/commit/a2fd710853d5b2e2b55bbc1a6540e9f8937a43fc>
> .
> [image: image]
> <https://user-images.githubusercontent.com/38635377/74680079-40547b00-5185-11ea-98f5-81a65f9210c2.png>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/navcoin/navcoin-core/pull/659?email_source=notifications&email_token=AAIDAKOUNVZZKJYH6LMYRYLRDLNBLA5CNFSM4KV6CSV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL7MKNQ#issuecomment-587121974>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDAKIWRNR2QZX4WQDC3VTRDLNBLANCNFSM4KV6CSVQ>
> .
>

navbuilder commented:
A new build of cc016af5d681717eeafa7aed588e09c495985e04 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

navbuilder commented:
A new build of a2fd710853d5b2e2b55bbc1a6540e9f8937a43fc has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

aguycalled commented on src/qt/navcoin.cpp:
```diff
@@ -333,13 +339,39 @@ void NavCoinApplication::createOptionsModel(bool resetSettings)
 
     pollShutdownTimer = new QTimer(window);
     connect(pollShutdownTimer, SIGNAL(timeout()), window, SLOT(detectShutdown()));
-    pollShutdownTimer->start(200);
```
what are the consequences of removing this?

mrmetech replied:
My bad I will add that back, when I was porting it, looked like I should remove it. It was moved in Upstream bitcoin so I guess I must have been a little confused or something when I did that, am gonna add it back now.

mrmetech replied:
Ok that changes is now reverted

navbuilder commented:
A new build of 3bd3c982d83e8ab5c29cdbc4dd445eea10c397ba has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui

aguycalled commented:
Tested MacOS, approved after last changes.

navbuilder commented:
A new build of 96a460a2096bb1e14758f25914106fc1a20485de has completed succesfully!
Binaries available at https://build.nav.community/binaries/mnemonic-startup-gui
